File: sdk/typescript/examples/balance.ts
/**
 * Read an on-chain ERC-1155 balance (litres of a batch held by an address).
 *
 *   npx tsx examples/balance.ts <address> <batchId>
 *
 * Pure read — no wallet, no funding needed. Routes to /read/ (cached LB).
 */
import { createReadClient, HaraPalmOil } from "@hara/registry-sdk";
import type { Address } from "viem";

async function main() {
  const account = (process.argv[2] ?? "0x0000000000000000000000000000000000000001") as Address;
  const batchId = BigInt(process.argv[3] ?? "1");

  const client = createReadClient();
  console.log("chainId:", await client.getChainId());
  console.log("block:  ", await client.getBlockNumber());

  const palmOil = new HaraPalmOil(client);
  const liters = await palmOil.balanceOf(account, batchId);
  console.log(`Balance of batch ${batchId} for ${account}: ${liters} litres`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sdk/python/examples/balance.py
"""Read on-chain state directly: chain height + an ERC-1155 batch balance.

Usage:
    python examples/balance.py <account_address> <batch_id>
"""

from __future__ import annotations

import sys

from hara_registry import HaraRegistry


def main() -> None:
    reg = HaraRegistry()

    print(f"chain id     : {reg.chain.get_chain_id()}")
    print(f"block number : {reg.chain.get_block_number()}")

    if len(sys.argv) < 3:
        print("\npass <account_address> <batch_id> to query a balance")
        return

    account = sys.argv[1]
    batch_id = int(sys.argv[2])

    litres = reg.palm_oil.balance_of(account, batch_id)
    print(f"\n{account} holds {litres} litres of batch {batch_id}")

    relay = reg.relay.address
    approved = reg.palm_oil.is_approved_for_all(account, relay)
    print(f"relay {relay} approved as operator: {approved}")


if __name__ == "__main__":
    main()
